### ISSUE TITLE
JCache bootstrap should have a way to reconciliate existing `HazelcastInstance`

### ISSUE THREAD

**Issue by snicoll** [hazelcast/hazelcast#9958]
This is an issue raised on the Spring Boot tracker for which I think there's nothing we can do. Spring Boot auto-configures hazelcast since it's not specific to caching. It turns you might use hazelcast for that as well.

If you chose to use JCache, then we bootstrap via the standard `JCache` API (obviously) and doing so leads to Hazelcast starting again. I think there should be a way for Hazelcast to figure out that an `HazelcastInstance` has already been started on this `ClassLoader` and that configuration URI and reuse that rather than creating a fresh one. As a user of the `JCache` API there's not much we can do.

We already have a [hazelcast specific hack](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/JCacheCacheConfiguration.java#L126) because Hazelcast' doesn't honour the URI parameter of the `CacheProvider#getCacheManager` properly.

You can easily reproduce using [this sample](https://github.com/snicoll-scratches/test-hazelcast-jcache): uncomment the `spring.cache.type` property in `application.properties` and you'll see two members starting in the cluster.

**Comment by vbekiaris** [comment]
Thanks @snicoll for raising this issue, identifying an existing or creating a new `HazelcastInstance` via `JCache` API could use some clean up to make it simpler and more predictable.

**Comment by dirkvanrensburg** [comment]
Just some more information:

**1**
If you set the `spring.cache.type` to only get one instance then Hazelcast does not use the `<cache>` configuration in `hazelcast.xml`. It creates a plain map instead using default config if no `<map>` config is defined in `hazelcast.xml`

**2**
If you allow it to run with two instances (by removing `spring.cache.type`) then the `<cache>` configuration is used, in fact Hazelcast fails to start if `<cache>` config is missing in this case:

    java.lang.IllegalArgumentException: Cannot find cache named 'numbers'
    for CacheResultOperation

**Comment by vbekiaris** [comment]
#9973 fixes `HazelcastInstance` resolution when constructing a `CacheManager` to allow use of an instance name or Hazelcast configuration file location as `URI`; this way we avoid having to use properties to indicate the instance name of an existing `HazelcastInstance` or the config file location that resolves `HazelcastInstance` with get-or-create semantics (as long as an instance name is specified in the config file).

The options to obtain a `CacheManager` while re-using an existing `HazelcastInstance` are:
* After having constructed a `HazelcastInstance`, obtain its name with `HazelcastInstance.getName()` and construct the `CacheManager` as described in [Binding to a named instance](http://docs.hazelcast.org/docs/3.8/manual/html-single/index.html#binding-to-a-named-instance). Example:
```
HazelcastInstance instance = Hazelcast.newHazelcastInstance();
CachingProvider cachingProvider = Caching.getCachingProvider();
CacheManager cacheManager = cachingProvider.getCacheManager(null, null, HazelcastCachingProvider.propertiesByInstanceName(instance.getName());
```
* Since Hazelcast 3.7, instead of instance name, you may pass the instance object itself as a property like this:
```
CacheManager cacheManager = cachingProvider.getCacheManager(null, null, HazelcastCachingProvider.propertiesByInstanceItself(instance);
```

Notice that in the examples above I used convenience methods, these simply create a `Properties` object and put the provided value in the appropriate property (`hazelcast.instance.name` or `hazelcast.instance.itself` respectively).

Unfortunately there is no way to bind a standalone `HazelcastInstance` to the `URI` & `ClassLoader` that identify a `CacheManager`, so binding `CacheManager` to an already started `HazelcastInstance` must be done explicitly via instance name (or via a `Config` object that specifies an instance name).

**Comment by snicoll** [comment]
@vbekiaris thanks for the details! I understand there is a limitation in Hazelcast that is going to be fixed with #9973 - I've added a comment there to make sure I understood things properly.

I will update Spring Boot so that the user can pass the instance name as a proprietary parameter. My initial test shows that fixes the issue for that user. I am really not keen to have `Hazelcast` API in our JCache auto-configuration so looking for an existing `HazelcastIntance` to get its name and/or instance is not an option.

**Comment by vbekiaris** [comment]
I think it sounds good! Ping me on gitter if you need any further information.
